### PR TITLE
jdk: keep source files for IDE support

### PR DESCRIPTION
**Pull request by yawnt**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

some IDEs need access to src.zip in order to provide better autocompletion

### CODE REVIEW

**Comment by cleverca22** [comment]
http://mywiki.wooledge.org/ParsingLs

**Comment by taku0** [comment]
How about `javafx-src.zip`?

File: pkgs/development/compilers/oraclejdk/jdk-linux-base.nix [hunk 1]
```diff
@@ -103,7 +103,7 @@ let result = stdenv.mkDerivation rec {
       mv $sourceRoot $out
     fi
 
-    for file in $out/*
+    for file in $(ls $out/* | grep -v "src.zip")
```
Comment (joachifm) [inline pkgs/development/compilers/oraclejdk/jdk-linux-base.nix]: I believe you can do
```
shopt -s extglob
for file in $out/!(src.zip)
```
to achieve the same with pure bash.

**Comment by yawnt** [comment]
@taku0 grep -v also filters out `javafx-src.zip`

File: pkgs/development/compilers/oraclejdk/jdk-linux-base.nix (hunk 1 above)
Comment (yawnt) [inline pkgs/development/compilers/oraclejdk/jdk-linux-base.nix]: this wouldn't filter out `javafx-src.zip` tho :/

**Comment by yawnt** [comment]
i've rebased according to @joachifm 's suggestion :)

File: pkgs/development/compilers/oraclejdk/jdk-linux-base.nix (hunk 1 above)
Comment (joachifm) [inline pkgs/development/compilers/oraclejdk/jdk-linux-base.nix]: The above is an example of an inverted glob, so `!(*src.zip)` should do what you want.

Comment (joachifm) [inline pkgs/development/compilers/oraclejdk/jdk-linux-base.nix]: Ah, I see that's what you did :)

**Comment by joachifm** [comment]
Please use a more descriptive commit subject, though, this will annoy someone slogging through the history late at night.

**Comment by yawnt** [comment]
done :+1:

**Comment by joachifm** [comment]
The desired format is described in the [contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md)

**Comment by copumpkin** [comment]
Could the sources live in a separate output? Or are they not very large?

**Comment by yawnt** [comment]
they're suppposed to stay in $JAVA_HOME, that's where tools look for them :/

**Comment by danbst** [comment]
add a bit of `lndir` magic (like in https://github.com/NixOS/nixpkgs/issues/15011#issuecomment-215151047), and you can have several outputs and `JAVA_HOME` pointing to their combination.

**Comment by yawnt** [comment]
@copumpkin they're 25MB combined, so pretty insignificant I'd say

@danbst where would the advantage be in that solution?

**Comment by danbst** [comment]
@yawnt other than reduced /nix/store size and finer dependencies, none.

@copumpkin @joachifm Maybe we should have nice syntax for joining outputs? For example:

```
let
wrapper = ''
  export JAVA_HOME="${pkgs.oraclejdk.combine [out dev]}"
  ...
'';
```

That would be better than 
```
let
wrapper = ''
  export JAVA_HOME="${pkgs.oraclejdk.override { includeSources = true; }"
  ...
'';
```
because doesn't trigger rebuild.

**Comment by danbst** [comment]
apacheAnt requires JDK but may not require sources

**Comment by joachifm** [comment]
I agree 25MB is not a lot.

If all IDEs/editors/whatever end up setting `JAVA_HOME` independently in their wrappers anyway, it does make some sense to do what @danbst is suggesting, though, so that only IDE/editor users need bear the cost. I don't know anything about java, though, so ultimately this is something y'all need to figure out.

**Comment by zimbatm** [comment]
Pardon my lack of understanding, what are the sources used for ? If the second output is pointing to the same folder then if both outputs are installed in the profile the $JAVA_HOME might point to the same folder ?

**Comment by joachifm** [comment]
@zimbatm based on the discussion at https://github.com/NixOS/nixpkgs/pull/13661, my understanding is that `JAVA_HOME` is typically set explicitly, and is not picked up from the user's profile.

**Comment by zimbatm** [comment]
It goes beyond this PR but I think we need a unified way to generate profile.d files that work on the system level, user level and per profile. There should be a way for `$profile/etc/profile.d/jre.sh` to set the JAVA_HOME to `$profile/lib/openjdk` so that it's stable on a profile-level. So when the source is installed it will be available in the same path.

**Comment by joachifm** [comment]
Merging this, please address any shortcomings in a future PR ...

**Pull request merged**